File: sorts.py
# Sample implementations of basic sorts in python
# The implementations are not aimed to be roboust, but
# to be instructive
import copy
import numpy

def isSorted(x):
	"""
	Return True if x is sorted
	"""
	for i in range(len(x)-1):
		if x[i] > x[i+1]:
			print i, i+1, x[i], x[i+1]
			return False

	return True

def swap(x,i,j):
	"""
	Swap the ith and the jth element of the array x
	"""
	temp = x[i]
	x[i] = x[j]
	x[j] = temp


def insertionSort(x):
	"""
	Sort the array-like object x inplace with insertion sort
	"""

	for i in range(len(x)):
		for j in range(i):
			if x[i-j] < x[i-j-1]:
				swap(x,i-j,i-j-1)
			else:
				break


def selectionSort(x):
	"""
	Sort the array-like object x inplace with selection sort
	"""
	for i in range(len(x)):
		minId = i
		for j in range(i,len(x)):
			minId = minId if x[minId] < x[j] else j
		swap(x,i,minId)

def mergeSort(x,**kwargs):
	"""
	Sort the array-like object x with merge sort. 
	"""
	start = kwargs.get('start',0)
	end = kwargs.get('end',len(x))

	if end - start == 1 or end == 0: # Array of length 1 is always sorted, also return for empty array
		return

	# Construct auxillary array for merging
	aux = kwargs.get('aux',copy.copy(x))

	# Partition the array and sort the parts
	mid = numpy.floor(0.5*(end+start)).astype('int')
	mergeSort(x,start=start,end=mid,aux=aux)
	mergeSort(x,start=mid,end=end,aux=aux)


	# Merge the sorted partitions into the aux array
	ptr1 = start
	ptr2 = mid
	for i in range(start,end):
		if ptr1 == mid:
			aux[i] = x[ptr2]
			ptr2 += 1
		elif ptr2 == end:
			aux[i] = x[ptr1]
			ptr1 +=1
		elif x[ptr1] < x[ptr2]:
			aux[i] = x[ptr1]
			ptr1+=1
		else:
			aux[i] = x[ptr2]
			ptr2+=1

	# Copy the aux array back to x.
	for i in range(start,end):
		x[i] = aux[i]


def heapSort(x):
	"""
	Sorts the array x by using heapsort
	"""
	h = heap(len(x),x)
	for i in range(len(x)):
		x[len(x)-i-1] = h.delMax()

	
class heap:
	def __init__(self,N,data=None):
		"""
		Initialize a heap of size N, with data such that len(data) <=N
		"""
		self.x = [None]	# Heap initially empty. There is no nice way to ensuring that x never grows beyond 
								# length N in python, so we will enforce this "manually"
		self.heapLen = 0
		self.maxHeapLen = N
		try:
			iterator = iter(data)
			if len(data) > N:
				raise Exception('Heap overflow. Too much data')
			else:
				[self.insert(p) for p in data]
		except TypeError:
			pass
		

	def insert(self,d):
		"""
		Insert data 'd' into the heap
		"""
		if self.heapLen == self.maxHeapLen:
			raise Exception('Heap overflow. Cannot insert in a full heap')

		self.x.append(d)
		self.heapLen += 1
		self.__swim()

	def delMax(self):
		if self.heapLen > 0:
			swap(self.x,1,self.heapLen)
			self.heapLen -= 1
			self.__sink()
			return self.x.pop()
		else:
			return None
	
	def delMin(self):
		if self.heapLen > 0:
			self.heapLen -= 1
			return self.x.pop()
		else:
			return None
	
	def __swim(self):
		"""
		Re-heapyfy with swim
		"""
		child = self.heapLen
		heaped = True if child <= 1 else False 	# heap of length 1 or less is always in heap order

		while not heaped and child > 1:
			parent = numpy.floor(0.5*child).astype(int)
			if self.x[child] > self.x[parent]:
				swap(self.x,child,parent)
				child = parent
			else:
				heaped = True


	def __sink(self):
		"""
		Re-heapyfy with sink
		"""
		parent = 1
		heaped = True if self.heapLen <= 1 else False # Heap of len 1 or less is always in heap order

		while not heaped:
			child1, child2 = parent*2, parent*2+1
			if child1 > self.heapLen:
				break
			elif child2 > self.heapLen:
				maxChild = child1
				heaped = True
			elif self.x[child1] > self.x[child2]:
				maxChild = child1
			else:
				maxChild = child2
			if self.x[parent] < self.x[maxChild]:
				swap(self.x,parent,maxChild)
				parent = maxChild
			else:
				heaped = True
				
def quicksort(x):
	"""
	Sort x inplace using the classical quicksort algorithm.
	We do not randomize; the user must.
	"""
	def partition(x,lo,hi):
		"""
		Partition the array x[lo,...,hi-1] such that the entry originally
		at x[lo] is at its correct sorted position. Return 'j' where j 
		is the correct index of this entry
		"""
		ptr1, ptr2 = lo+1, hi
		partitioned = False

		while not partitioned:
			while x[ptr1] <= x[lo] and ptr1 < hi:
				ptr1 += 1
		
			while x[ptr2] > x[lo] and ptr2 > lo:
				ptr2 -= 1

			if ptr1 >= ptr2 or ptr1 == hi or ptr2 == lo:
				partitioned = True
			else:
				swap(x,ptr1,ptr2)

		swap(x,lo,ptr2)

		return ptr2

	def sort(x,lo,hi):
		if hi <= lo:
			return
		pivot = partition(x,lo,hi)
		sort(x,lo,pivot-1)
		sort(x,pivot+1,hi)

	sort(x,0,len(x)-1)

File: sorts.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

typedef enum {true, false} myBool;
#define arrayPrintPerLine 5

myBool isSorted(int *x, int len)
{
	/* 
	Return true if array x of length len is sorted. 
	*/

	myBool sorted = true;
	int count = 0;
	for(count = 0; count < len-1; count++)
	{
		if (x[count] > x[count+1])
		{
			printf("Breaking sort order with x[%d] = %d, x[%d] = %d\n",count,x[count],count+1,x[count+1]);
			sorted = false;
			break;
		}
	}

	return sorted;
}


int * randomArray(int len)
{
	// Random integer array of lenght len
	int *x = (int *)malloc(len*sizeof(int));
	srand(time(NULL));
	while(len > 0)
	{
		x[--len] = rand();
	}
	return x;
}

void printArray(int *x, int len)
{
	
	int count = 0;
	for(count = 0; count < len; count++)
	{
		printf("x[%.3d] = %.10d\t",count,x[count]);
		if ((count+1)%arrayPrintPerLine == 0) printf("\n");
	}

	printf("\n");
}

void assertSorted(int *x, int len, char *name)
{
	if (isSorted(x,len) == true)
	{
		printf("%s is sorted\n",name);
	}
	else
	{
		printf("%s is unsorted\n",name);
	}
}

void swap(int *x, int i, int j)
{
	int temp = x[i];
	x[i] = x[j];
	x[j] = temp;
}

void insertionSort(int *x, int len)
{
	int i, j;
	for(i = 0; i < len; i++)
	{
		for(j=0; j<i; j++)
		{
			if(x[i-j]<x[i-j-1])
			{
				swap(x,i-j,i-j-1);
			}
			else
			{
				break;
			}
		}
	}
}

void selectionSort(int *x, int len)
{
	int i, j;
	
	for(i=0; i<len; i++)
	{
		int minInd = i;
		for(j=i; j<len; j++)
		{
			minInd = x[j] < x[minInd] ? j : minInd;
		}
		swap(x,i,minInd);
	}
}

void merge(int *x, int len, int div)
{
	/*
	Merge the array x of length 'len' into 
	one sorted array, if x[0...div-1], and 
	x[div... len-1] are sorted
	*/

	// auxillary array for sorting
	int *aux = (int *)malloc(len*sizeof(int));

	// pointers to walk the two divisions
	int ptr1 = 0, ptr2 = div, auxPos = 0;

	while(ptr1 < div || ptr2 < len)
	{
		if (ptr1 < div && ptr2 < len)
		{
			// None of the divisions is exhausted, its business as usual
			if (x[ptr1] < x[ptr2])
			{
				aux[ptr1 + ptr2 - div] = x[ptr1];
				ptr1 += 1;
			}
			else
			{
				aux[ptr1 + ptr2 - div] = x[ptr2];
				ptr2 += 1;
			}
		}
		else if(ptr1 < div)
		{
			aux[ptr1 + ptr2 - div] = x[ptr1];
			ptr1 +=1;
		}
		else
		{
			aux[ptr1 + ptr2 - div] = x[ptr2];
			ptr2 +=1;
		}
	}

	while(len > 0)
	{
		x[len-1] = aux[len-1];
		len -= 1;
	}

	free(aux);
}

void mergeSort(int *x, int len)
{
	/*
	Mergesort an array of length len
	*/
	if (len < 2) return;	// array of length 1 or 0 is always sorted.
	
	int div = len/2;	// Divide array into left and right bits. 
										// left: 0,...,div-1 right: div,...,len-1

	mergeSort(x,div);
	mergeSort(x+div,len-div);
	merge(x,len,div);
}


// **** START HEAP FUNCTIONS ETC. **** //
typedef struct
{
	// An array based datastructure to store a binary heap.
	int maxSize;
	int size;
	int *x;
} heap;

heap * heap_initialize(int maxSize)
{
	heap *h = (heap *)malloc(sizeof(heap));
	h->maxSize = maxSize;
	h->x = (int *)malloc((maxSize+1)*sizeof(int));
	h->size = 0;
	return h;
}

void heap_swim(heap *h)
{
	/*
	Re-establish heap order by "swimming" the last 
	element to its correct position
	*/

	int swimmer = h->size;

	while (swimmer != 1)
	{
		int parent = (int) floor(swimmer*0.5);
		if (h->x[parent] < h->x[swimmer])
		{
			swap(h->x,parent,swimmer);
			swimmer = parent;
		}
		else
		{
			break;
		}
	}

}

void heap_sink(heap *h)
{
	/* 
	Re-establish heap order by sinking the first 
	element to its correct position.
	*/

	int sinker = 1;
	myBool heaped = false;

	while(heaped == false)
	{
		int lChild = 2*sinker, rChild = 2*sinker + 1, maxChild=sinker;

		if(lChild > h->size)
		{ 
			heaped = true;
		}
		else if(rChild > h->size) 
		{ 
			heaped = true; 
			maxChild = h->x[lChild] > h->x[sinker] ? lChild : sinker;
		}
		else if(h->x[rChild] > h->x[sinker] || h->x[lChild] > h->x[sinker])
		{
			maxChild = h->x[rChild] > h->x[lChild] ? rChild : lChild;
		}
		else
		{
			heaped = true;
		}
		swap(h->x,maxChild,sinker);
		sinker = maxChild;
	}
}

void heap_print(heap *h)
{
	printArray(h->x+1,h->size);
}

void heap_push(heap *h, int data)
{
	// Check for heap overflow
	if (h->size == h->maxSize)
	{
		printf(" Trying to push to full heap. Exiting\n");
		fflush(stdout);
		exit(EXIT_FAILURE);
	}

	// Add data to heap
	h->size += 1;
	h->x[h->size] = data;

	// Re-establish heap order
	heap_swim(h);
}

int heap_pop(heap *h)
{
	int popped = h->x[1];

	swap(h->x,1,h->size);
	h->size -= 1;
	
	heap_sink(h);

	return popped;
}

void heap_free(heap **h)
{
	free((*h)->x);
	free(*h);
}

void heap_sort(int *x, int len)
{
	/*
	Sort x with a heap sort. 
	*/

	heap *h = heap_initialize(len);

	int count = 0;
	for(count = 0; count < len; count++)
	{
		heap_push(h,x[count]);
	}


	for(count = 0; count < len; count++)
	{
		x[len-count-1] = heap_pop(h);
	}

	heap_free(&h);
}

// **** END HEAP FUNCTIONS ETC. **** //

// **** START PRIORITY QUEQE FUNCTIONS ETC. **** //
typedef struct 
{
	int maxSize, size;
	heap *h;
} priorityQ;

priorityQ * priorityQ_initialize(int maxSize)
{
	priorityQ * pQ = (priorityQ *)malloc(sizeof(priorityQ));
	pQ->maxSize = maxSize;
	pQ->size = 0;
	pQ->h = heap_initialize(maxSize);

	return pQ;
}

void priorityQ_insert(priorityQ *pQ, int data)
{
	if(pQ->size < pQ->maxSize)
	{
		pQ->size += 1;
		heap_push(pQ->h, data);	
	}
	else if (data < (pQ->h)->x[1])
	{
		heap_pop(pQ->h);
		heap_push(pQ->h,data);
	}
}

int priorityQ_pop(priorityQ *pQ)
{
	if (pQ->size == 0)
	{
		printf("Popping empty priorityQ. Exiting\n");
		exit(EXIT_FAILURE);
	}

	pQ->size -= 1;
	return heap_pop(pQ->h);
}

void priorityQ_print(priorityQ *pQ)
{
	heap_print(pQ->h);
}

// **** END PRIORITY QUEQE FUNCTIONS ETC. **** //


int qs_partition(int *x, int lo, int hi)
{
	int ptr1 = lo+1, ptr2 = hi;
	myBool partitioned = false;

	while (partitioned == false)
	{
		while (x[ptr1] < x[lo] && ptr1 < hi)
		{
			ptr1 += 1;
		}

		while (x[ptr2] > x[lo] && ptr2 > lo)
		{
			ptr2 -= 1;
		}

		if (ptr1 >= ptr2 || ptr1 == hi || ptr2 == lo)
		{
			partitioned = true;
		}
		else
		{
			swap(x,ptr1,ptr2);
		}
	}

	swap(x,lo,ptr2);
	return ptr2;
}

void qs_sort(int *x, int lo, int hi)
{
	if (hi <= lo)	return;

	int pivot = qs_partition(x,lo,hi);
	qs_sort(x,lo,pivot-1);
	qs_sort(x,pivot+1,hi);
}

int main()
{
	int len = 10000000;
	char *name = "X";
	int *x;
	//printArray(x,len);

	/*
	x = randomArray(len);
	printf("\nSorting %s with insertion sort\n",name);
	insertionSort(x,len);
	//printArray(x,len);
	assertSorted(x,len,name);
	free(x);


	x = randomArray(len);
	printf("\nSorting %s with selection sort\n",name);
	selectionSort(x,len);
	//printArray(x,len);
	assertSorted(x,len,name);
	free(x);
	x = randomArray(len);
	printf("\nSorting %s with merge sort\n",name);
	mergeSort(x,len);
	//printArray(x,len);
	assertSorted(x,len,name);
	free(x);

	x = randomArray(len);
	printf("\nSorting %s with heap sort\n",name);
	heap_sort(x,len);
	assertSorted(x,len,name);
	free(x);

	x = randomArray(len);
	printf("\nSorting %s with quick sort\n",name);
	qs_sort(x,0,len-1);
	assertSorted(x,len,name);
	free(x);
	*/
	
	
	x = randomArray(len);
	int pQSize = 10;
	priorityQ *pQ = priorityQ_initialize(pQSize);
	{
		int count = 0;
		for (count = 0; count < len; count++)
		{
			priorityQ_insert(pQ,x[count]);
		}
	}
	while(pQ->size > 0)
	{
		printf("%d\n",priorityQ_pop(pQ));
	}

	mergeSort(x,len);
	printArray(x,pQSize);

	free(x);

	return 1;
}
